JoomGalleryfriends/JG4-dev
# Config service

## What is this service for

The Config service will be used to handle the different configurations set in the component. Based on the chosen inheritance method in the Glocal component settings, the correct set of configurations are loaded depending on the current context. The following configuration sets are taken into account:

- Global Configuration set (Config manager)
- User Configuartion set (Config manager)
- Category settings (Category edit form)
- Image settings (Image edit form)
- Menu item settings (Menu item edit form)

The following contexts are available:

- com_joomgallery (General settings)
- com_joomgallery.user.id (Settings for a specified user)
- com_joomgallery.category.id (Settings for all images in a specified category)
- com_joomgaller.image.id (Setting for a specified image)
- com_joomgallery.menu.id (Settings for a specified menu item)

## How does it work
When you instantiate the service, you have to provide a context and optionally an ID telling the service for which part of the component you want to have the configuration set loaded.
The service then will get the configuration sets and override the settings based on the chosen inheritance method. Default inheritance method will override the settings as follows:

1. Global Configuration set
2. User Configuartion set
3. Category settings (taking into account all parent categories)
4. Image settings
5. Menu item settings (frontend only, lookign which site are you accessing)

## How to use this service

You can load the service anywhere in Joomla by doing

```
$component = Factory::getApplication()->bootComponent('com_joomgallery');  // boot JoomGallery component
$component->createConfig('context', ID);   // instantiate service based on the provided context
$config = $component->getConfig();   // load the service into your scope
$config->get('jg_param'); // get parameter with key name 'jg_param'
```

## How to test the service

Make sure "Debug System" is "Yes" in Joomla Global configuration.
Open the file `/administrator/com_joomgallery/src/View/Faulties/HtmlView.php` and insert the following code on line 62 after `throw new \Exception(implode("\n", $errors)); }`

Different test instructions see comments below...
